src/openzeppelin/token/erc721.cairo

```diff
@@ -0,0 +1,350 @@
+        ERC721::approve(to, token_id)
+    }
+
+    #[external]
+    fn set_approval_for_all(operator: ContractAddress, approved: bool) {
+        ERC721::set_approval_for_all(operator, approved)
+    }
+
+    #[external]
+    fn transfer_from(from: ContractAddress, to: ContractAddress, token_id: u256) {
+        ERC721::transfer_from(from, to, token_id)
+    }
+
+    #[external]
+    fn safe_transfer_from(
+        from: ContractAddress, to: ContractAddress, token_id: u256, data: Array<felt252>
+    ) {
+        ERC721::safe_transfer_from(from, to, token_id, data)
+    }
+
+    #[internal]
+    fn initializer(name_: felt252, symbol_: felt252) {
+        _name::write(name_);
+        _symbol::write(symbol_);
+        erc165::ERC165Contract::register_interface(erc721::IERC721_ID);
+        erc165::ERC165Contract::register_interface(erc721::IERC721METADATA_ID);
+    }
+
+    #[internal]
+    fn _owner_of(token_id: u256) -> Option<ContractAddress> {
+        let owner = _owners::read(token_id);
+        match owner.is_zero() {
+            bool::False(()) => Option::Some(owner),
+            bool::True(()) => Option::None(())
+        }
+    }
+
+    #[internal]
+    fn _exists(token_id: u256) -> bool {
+        !_owners::read(token_id).is_zero()
+    }
+
+    #[internal]
+    fn _is_approved_or_owner(spender: ContractAddress, token_id: u256) -> bool {
+        let owner = _owner_of(token_id).expect('ERC721: invalid token ID');
+        owner == spender | is_approved_for_all(owner, spender) | spender == get_approved(token_id)
+    }
+
+    #[internal]
+    fn _approve(to: ContractAddress, token_id: u256) {
+        let owner = _owner_of(token_id).expect('ERC721: invalid token ID');
+        assert(owner != to, 'ERC721: approval to owner');
+        _token_approvals::write(token_id, to);
+        Approval(owner, to, token_id);
+    }
+
+    #[internal]
+    fn _set_approval_for_all(owner: ContractAddress, operator: ContractAddress, approved: bool) {
+        assert(owner != operator, 'ERC721: self approval');
+        _operator_approvals::write((owner, operator), approved);
+        ApprovalForAll(owner, operator, approved);
+    }
+
+    #[internal]
+    fn _mint(to: ContractAddress, token_id: u256) {
+        assert(!to.is_zero(), 'ERC721: invalid receiver');
+        assert(!_exists(token_id), 'ERC721: token already minted');
+
+        // Update balances
+        _balances::write(to, _balances::read(to) + 1.into());
+
+        // Update token_id owner
+        _owners::write(token_id, to);
+
+        // Emit event
+        Transfer(Zeroable::zero(), to, token_id);
+    }
+
+    #[internal]
+    fn _transfer(from: ContractAddress, to: ContractAddress, token_id: u256) {
+        assert(!to.is_zero(), 'ERC721: invalid receiver');
+        let owner = _owner_of(token_id).expect('ERC721: invalid token ID');
+        assert(from == owner, 'ERC721: wrong sender');
+
+        // Implicit clear approvals, no need to emit an event
```

Comment:
I think you do need to emit an event here. The EIP spec mentions "This emits when the approved address for an NFT is changed or reaffirmed." for `Approval`. FWIW, the Cairo 0 version does emit an event, via a call to `_approve`.

Reply:
In solidity we decided to remove it. Other libraries such as solmate also [don't emit an event here](https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC721.sol#L82-L109)

Reply:
I didn't miss that part, I just don't think emitting here contradicts it, same as you argued in the linked discussion (thanks for pointing me to it, interesting read). Regardless, I understand that you want to keep in line with the Solidity implementations, that's fair 👍